[Feature] 이메일 인증코드 전송, 검증 기능을 구현한다.

ghrltjdtprbs opened the pull request:
# ⭐️ [Feature] 이메일 인증코드 전송, 검증 기능을 구현한다.

- 이메일 인증 코드를 전송, 검증하는 기능을 구현하였습니다.
- 리소스 절약을 위해 local, test 환경에서는 실제 이메일이 발송되지 않고, 로그로 찍힙니다.
- 배포 환경(develop)에서는 실제 이메일이 발송됩니다.  
- 발송된 인증 코드의 유효시간은 하루입니다. 

## 🛠️ 변경 사항

- POST members/email : 인증코드 전송
- POST members/email/verification : 인증코드 검증

## ❗️ 관련 이슈

- #58 

## ❗️ 개발 유형

- [ ] 버그 수정
- [x] 새로운 기능 개발
- [ ] 리팩토링
- [ ] 테스트 코드 작성
- [ ] 문서 업데이트

## ⏱️ 작업 기간

- 작업 시작일: (2024-05-16)
- 작업 종료일: (2024-05-17)

## 💡 유의사항

- 질문이나 개선 사항이 있다면 코드리뷰 남겨주세요!
- 이메일 템플릿 디자인은 간단하게 제작하였습니다. 프론트나 UIUX 분들이 괜찮다고 하면 그대로 사용할 예정입니다.  
 
<img width="813" alt="스크린샷 2024-05-17 03 06 11" src="https://github.com/10aeat/10aeat_Backend/assets/105612931/b84c4bb2-c98d-47e9-b6ab-48acb7aeabe2">

ghrltjdtprbs committed:
[feat] JavaMailSender 의존성 추가

변경 사항 :
 - build.gradle에 JavaMailSender의존성을 추가하였습니다.
 - Gson 라이브러리 의존성 추가 (버전 2.8.9)
 - validation 의존성 추가

ghrltjdtprbs committed:
[feat] 이메일 로직 예외 클래스 추가

변경 사항 :
 - ErrorCode 추가
 - EmailSendingException, EmailTemplateLoadException, VerificationCodeExpiredException 예외 클래스 추가

ghrltjdtprbs committed:
[feat] EmailUseCase 클래스 추가

문제 사항 :
 - local,test 환경에서는 테스트시 실제 이메일을 전송할 필요 없음(리소스 낭비)
 - local,test 환경과 develop(배포) 환경의 이메일 서비스 클래스를 분리하기 위해 EmailUseCase를 추가하였습니다.

변경 사항 :
 - LocalEmailService와 DevelopEmailService는 공통된 로직을 사용하기 때문에 더 효율적인 코드 관리를 위해 EmailUseCase를 추가

ghrltjdtprbs committed:
[feat] 이메일 로직에 사용될 dto 클래스 추가

변경 사항 :
 - 이메일 인증번호 발송에 requestbody로 쓰일 EmailRequestDto 클래스 추가
 - 인증번호 검증 requestbody로 쓰일 EmailVerificationRequestDto 클래스 추가
 - 인증번호 검증 responsebody로 쓰일 EmailVerificationResponseDto추가.
 - EmailVerificationResponseDto 클래스는 record가 아닌 일반 dto 클래스로 작성하였습니다.
    이유 :
    record 타입의 모든 필드는 기본적으로 final로 선언되기 때문에, Gson과 같은 JSON 라이브러리가 리플렉션을 통해 이러한 필드들에 값을 할당하려 할 때 IllegalAccessException이 발생할 수 있음.

ghrltjdtprbs committed:
[feat] 이메일 템플릿 추가

변경 사항 :
 - 인증코드 전송시 이메일 양식으로 쓰일 email_template.html 파일을 추가하였습니다.

ghrltjdtprbs committed:
[refactor] 모든 타입 반환이 가능하도록 ResponseDTO 수정

변경 사항 :
 - 반환 타입을 Void -> T 로 수정

ghrltjdtprbs committed:
[feat] 이메일 설정을 위한 EmailConfig 클래스 추가

변경 사항 :
 - 실제 이메일을 전송하기 위해 이메일 설정을 관리하는 EmailConfig를 추가하였습니다.

ghrltjdtprbs committed:
[feat] EmailService 클래스 추가

변경 사항 :
 - DevelopEmailService: 배포 환경에서 실제 이메일 전송을 처리합니다.
 - LocalEmailService: 로컬 및 테스트 환경에서 이메일을 로그로 대체합니다.

ghrltjdtprbs committed:
[feat] application-develop.yml에 환경변수 추가

변경 사항 :
 - 배포환경에서 환경 변수를 git action secrets으로 관리하기 위해 메일 관련 환경 변수를 추가하였습니다.

ghrltjdtprbs committed:
[style] 공백 제거

변경 사항 :
 - application-local.yml의 불필요한 공백 제거

ghrltjdtprbs committed:
[feat] EmailController 클래스 추가

변경 사항 :
 - POST members/email : 인증코드 전송
 - POST members/email/verification : 인증코드 검증

ghrltjdtprbs committed:
[refactor] ResponseDTO 클래스 수정

문제 사항 :
 - 실제 응답값은 null필드를 출력하지 않지만, REST Docs 문서화시 null필드를 문서화 해야하는 문제점 발생

변경 사항 :
 - @JsonInclude(JsonInclude.Include.NON_NULL) 어노테이션을 추가
 - 기존에도 null 데이터는 응답값에 포함되지 않으나, RestDocs 문서화 시 null 필드로 인한 예외를 방지하기 위해 추가한 코드입니다.

ghrltjdtprbs committed:
[feat] lombok관련 의존성 추가

문제 사항 :
 - 테스트 코드에서 lombok 어노테이션 사용이 불가능한 문제 발생

변경 사항 :
 - 관련 의존성을 추가하였습니다.

ghrltjdtprbs committed:
[feat] 공통 응답값 REST Docs문서화

변경 사항 :
 - 공통 응답을 문서화하기 위한 관련 테스트 코드를 추가하였습니다.
 - CommonDocsController, CommonDocsTest 추가

ghrltjdtprbs committed:
[feat] RestDocs 테스트 지원을 위한 RestDocsSupport 추가

변경 사항 :
 - RestDocs 문서화를 위한 기본 설정을 제공하는 추상 클래스 RestDocsSupport를 추가하였습니다.
 - MockMvc 설정과 ObjectMapper 설정을 포함하여 각 테스트 클래스에서 일관된 RestDocs 설정을 사용할 수 있도록 하였습니다.
 - CharacterEncodingFilter를 추가하여 UTF-8 인코딩을 보장하였습니다.
 - JavaTimeModule을 등록하여 LocalDateTime과 같은 Java 8 날짜 및 시간 API 타입을 지원하도록 하였습니다.

ghrltjdtprbs committed:
[feat] EmailController RestDocs 문서화 테스트 코드 추가

변경 사항 :
 - EmailController를 RestDocs로 문서화하기 위한 테스트 코드를 추가하였습니다.

ghrltjdtprbs committed:
[feat] Asciidoctor API 문서 템플릿 파일 추가

변경 사항 :
 - 10aeat API 문서를 위한 Asciidoctor 템플릿 파일을 추가하였습니다.
 - 이메일 API 링크 및 공통 응답 형식을 포함하였습니다.
 - HTTP 상태 코드와 API 공통 응답 요소를 표 형식으로 정리하였습니다.

ghrltjdtprbs committed:
[feat] 이메일 REST API 문서 파일 추가

변경 사항 :
 - 이메일 인증 코드 전송 및 검증에 대한 API 문서를 추가하였습니다.

ghrltjdtprbs committed:
[test] LocalEmailServiceTest 테스트 추가

변경 사항 :
 - LocalEmailService클래스를 테스트 하는 유닛 테스트 코드를 추가하였습니다.

Galmaeki commented on src/main/java/com/final_10aeat/domain/member/service/LocalEmailService.java:
```diff
@@ -0,0 +1,77 @@
+public class LocalEmailService implements EmailUseCase {
+
+    private final RedisTemplate<String, String> redisTemplate;
+    private final Gson gson = new Gson();
```
Spring에 내장된 Jackson이 아니라 Gson을 사용하신 이유가 따로 있나요?

ghrltjdtprbs replied:
데이터를 단순하게 레디스에 저장하고 다시 불러오는 작업이기도 하고, 가독성과 간편함을 위해 직렬화, 역직렬화 작업을 빠르게 수행할 수 있는 Gson을 채택하였습니다!

Galmaeki commented on src/main/java/com/final_10aeat/domain/member/service/DevelopEmailService.java:
```diff
@@ -0,0 +1,128 @@
+    }
+
+    @Override
+    public ResponseDTO<EmailVerificationResponseDto> verifyEmailCode(String email, String code) {
```
ResponseDto는 비즈니스 로직에 포함되는 클래스가 아니라서 컨트롤러로 올려도 될 것 같습니다!

ghrltjdtprbs replied:
지적 감사합니다! 놓친 부분이 있었네요 ㅎㅎ 수정해서 커밋했으니 확인 부탁드립니다 😊😊
커밋 : a82994fb8b4cd2975702b6069ca226ff20cd79f7

Galmaeki commented on src/main/java/com/final_10aeat/domain/member/controller/EmailController.java:
```diff
@@ -0,0 +1,37 @@
+    }
+
+    @PostMapping("/verification")
+    public ResponseEntity<ResponseDTO<?>> verifyEmail(
```
Interface에선 EmailVerificationResponseDto로 제네릭을 사용하셨는데 여기서 와일드 카드를 사용하신 이유가 뭔가요?

ghrltjdtprbs replied:
이 부분은 에러 해결하는 과정에서 임시로 추가했던 부분인데, 수정한다는 것을 깜빡했습니다..ㅎㅎ 변경해서 커밋했으니 확인해주세요!
커밋 : 9102a529344687c7ab1e46fc32bf0c97c2db2494

Galmaeki commented on src/main/java/com/final_10aeat/domain/member/dto/request/EmailRequestDto.java:
```diff
@@ -0,0 +1,19 @@
+    String email,
+    @NotNull(message = "가입자가 소유자인지 입주자인지 선택해주세요.")
+    MemberRole role,
+    String dong,
```
동은 필수 정보가 아닌걸로 가는 건가요?

ghrltjdtprbs replied:
office에는 다양한 형태의 건물이 있고, 실제 동이 없고 호만 가지고 있는 경우도 있기 때문에 dong을 필수 정보로 하기에는 무리가 있다 판단하였습니다! 😊😊

Galmaeki commented on src/main/java/com/final_10aeat/domain/member/dto/response/EmailVerificationResponseDto.java:
```diff
@@ -0,0 +1,18 @@
+@Builder
+@AllArgsConstructor
+@NoArgsConstructor
+public class EmailVerificationResponseDto {
```
레코드가 아닌 이유가 궁금합니다!

ghrltjdtprbs replied:
record 타입의 모든 필드는 기본적으로 final로 선언되기 때문에, Gson과 같은 JSON 라이브러리가 리플렉션을 통해 이러한 필드들에 값을 할당하려 할 때 IllegalAccessException이 발생할 수 있습니다! 🫠 
실제로 record를 사용하니 위의 문제가 발생하였고, 해결을 위해서는 불필요한 로직이 너무 많이 추가되어 변경하였습니다

Galmaeki reviewed (changes requested):
고생하셨습니다!
코멘트 남겼으니 확인 부탁드립니다!

asd42270 commented on src/main/java/com/final_10aeat/domain/member/service/DevelopEmailService.java:
```diff
@@ -0,0 +1,128 @@
+    }
+
+    @Override
+    public ResponseDTO<EmailVerificationResponseDto> verifyEmailCode(String email, String code) {
```
엇, 요거는 ResponseDto만 리턴해줘도 괜찮지 않을까요????

ghrltjdtprbs replied:
지적 감사합니다! 놓쳤던 부분이였어요 ㅎㅎ 수정해서 커밋해두었으니 확인 부탁드려요 👍👍
커밋 : a82994fb8b4cd2975702b6069ca226ff20cd79f7

asd42270 commented on src/main/java/com/final_10aeat/domain/member/service/EmailUseCase.java:
혹시 UseCase에 대해서 설명해 주실 수 있을까요??? 
유스케이스라 하면 UML에서의 유스케이스 밖에 들어본 적이 없어서요...!!

ghrltjdtprbs replied:
이 부분이 UseCase패턴에 따라 설계되어 클래스 이름을 EmailUseCase로 설정하였습니다!
 
- Use Case 패턴 : 소프트웨어 시스템의 기능을 개별적인 시나리오나 작업 단위로 나누어, 각 기능을 명확하게 정의하고 캡슐화하는 설계 방식

asd42270 reviewed (changes requested):
고생하셨습니다!!!

ghrltjdtprbs committed:
[refactor] EmailController verifyEmail 메서드 제네릭 타입 변경

변경 사항 :
 - 메서드 반환 타입을 ResponseDTO<?>에서 ResponseDTO<EmailVerificationResponseDto>로 변경하여 타입 안전성 및 코드 가독성 향상

ghrltjdtprbs committed:
Merge pull request #61 from 10aeat/feature/59

[Feature] Spring REST Docs 기본 세팅 및 이메일 로직 테스트 코드 추가

ghrltjdtprbs committed:
[refactor] 테스트코드 ci 활성화

변경 사항 :
 - Dockerfile에 주석처리 되어있던 "#RUN ./gradlew test" 를 활성화 하였습니다.

ghrltjdtprbs committed:
[feat] 인증번호 불일치 예외 클래스 추가

변경 사항 :
 - 인증번호가 불일치할 때 사용되는 예외클래스 InvalidVerificationCodeException를 추가하였습니다.
 - 관련 메시지를 ErrorCode에 추가하였습니다.

ghrltjdtprbs committed:
[feat] 서비스 계층의 verifyEmailCode 메서드가 비즈니스 로직만 처리하도록 수정

문제 사항 :
 - 서비스 계층에서 ResponseDTO를 반환하여 컨트롤러 계층과의 책임 분리가 명확하지 않음

변경 사항 :
 - 서비스 계층에서 ResponseDTO를 반환하지 않고 EmailVerificationResponseDto를 반환하도록 변경
 - 컨트롤러에서 ResponseDTO로 감싸서 반환하도록 수정하여 책임 분리 및 코드 가독성 향상

ghrltjdtprbs committed:
[test] VerifyEmailCode() 변경 부분 테스트코드 반영

문제 사항 :
 - VerifyEmailCode() 메서드에 변경 사항이 생겨 관련 테스트 코드가 실패

변경 사항 :
 - 변경사항 반영하여 테스트 코드 수정하였습니다.

ghrltjdtprbs committed:
Merge pull request #64 from 10aeat/featrue/58

[Refacor] 코드 리뷰 반영해 이메일 로직 리팩토링

Galmaeki reviewed (approved):
고생 많으셨습니다!

asd42270 reviewed (approved):
고생 많으셨습니다!!!
